### PR TITLE
Task.04 and Task.05

### PR DESCRIPTION
**Pull request by ziod**
Signed-off-by: Oleksandr Aleksieiev <oleksandr.aleksieiev@globallogic.com>

### CODE REVIEW

File: 05_timers/time_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/*
+ * COPYRIGHT (C) 2021 Oleksandr Aleksieiev
+ * all rights reserved
+*/
+#include <iomanip>
+#include "time_manager.h"
+
+namespace time_manager {
+    const std::map<const clockid_t, std::string> TimeManager::validClockIdsMap = {
```
Comment (serhii-perederii) [inline 05_timers/time_manager.cpp]: Why not const std::string&

File: 05_timers/time_manager.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        friend std::ostream& operator<<(std::ostream& out, const TimeManager& t);
+
+    private:
+        clockid_t clock_type_;
+        timespec res_{};
+        timespec tp_{};
+        static const std::map<const clockid_t, std::string> validClockIdsMap;
+
+        static void reconfigureTimeManager(TimeManager *m, clockid_t const &type);
```
Comment (serhii-perederii) [inline 05_timers/time_manager.h]: Is raw pointer necessary for this case? Maybe reference?

File: 05_timers/kernelspace/Makefile [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+obj-m += custom_timer.o
+
+.PHONY: clean check
+
+all: check
+	make -C ${KERNELDIR} M=$(PWD) modules
+clean:
+	make -C ${KERNELDIR} M=$(PWD) clean
+check:
+	${KERNELDIR}scripts/checkpatch.pl -f ./custom_timer.c
```
Comment (serhii-perederii) [inline 05_timers/kernelspace/Makefile:12]: Newline?

**Review by serhii-perederii — CHANGES_REQUESTED** [review]

File: 05_timers/time_manager.cpp (hunk 1 above)
Comment (ziod) [reply 05_timers/time_manager.cpp]: What for?

File: 05_timers/time_manager.h (hunk 2 above)
Comment (ziod) [reply 05_timers/time_manager.h]: May be. But what the difference on how you transfer the data to the method in that particles cular case?

File: 05_timers/kernelspace/Makefile (hunk 3 above)
Comment (ziod) [reply 05_timers/kernelspace/Makefile:12]: No, golithub will put small Red icon if you miss smth.

File: 05_timers/kernelspace/custom_timer.c [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+#include <linux/module.h> // Core header for loading LKMs into the kernel
+#include <linux/kernel.h> // Contains types, macros, functions for the kernel
+#include <linux/ktime.h> // Time operations purpose
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Oleksandr Aleksieiev <oleksandr.aleksieiev@globallogic.com>");
+MODULE_DESCRIPTION("Print delta time between reading attribute.");
+MODULE_VERSION("0.1");
+
+static const char *const error_codes[] = { "OK",
```
[the remark this answers is not in the document]
Comment (ziod) [reply 05_timers/kernelspace/custom_timer.c:15]: Ok

File: 05_timers/time_manager.cpp (hunk 1 above)
Comment (serhii-perederii) [reply 05_timers/time_manager.cpp]: Pair will pass a const reference, this is already solved by STL

**Comment by serhii-perederii** [comment]
Commits are ok

**Review by serhii-perederii — APPROVED** [review]

**Pull request merged**